crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/agent_info_update.rs

```diff
@@ -0,0 +1,309 @@
+                        error!(failed_to_update_agent_info_for_space = ?e);
+                    }
+                }
+            }
+
+            info!("AgentInfoUpdateTask finished");
+            task_this.write().is_finished = true;
+        });
+
+        this
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::AgentInfoUpdateTask;
+    use crate::actor::BroadcastData;
+    use crate::dht_arc::DhtArc;
+    use crate::event::PutAgentInfoSignedEvt;
+    use crate::spawn::actor::space::{
+        KAgent, KBasis, KSpace, MaybeDelegate, OpHashList, Payload, SpaceInternal,
+        SpaceInternalHandler, SpaceInternalHandlerResult, VecMXM, WireConHnd,
+    };
+    use crate::spawn::meta_net::MetaNetCon;
+    use crate::wire::Wire;
+    use crate::{GossipModuleType, KitsuneP2pError};
+    use futures::FutureExt;
+    use ghost_actor::actor_builder::GhostActorBuilder;
+    use ghost_actor::{
+        GhostControlHandler, GhostControlSender, GhostError, GhostHandler, GhostSender,
+    };
+    use kitsune_p2p_fetch::FetchContext;
+    use kitsune_p2p_types::agent_info::AgentInfoSigned;
+    use kitsune_p2p_types::KOpHash;
+    use parking_lot::RwLock;
+    use std::collections::HashSet;
+    use std::sync::Arc;
+    use std::time::{Duration, Instant};
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn update_agent_info() {
+        let (test_sender, _) = setup(DummySpaceInternalImpl::new()).await;
+
+        // It should be possible to set this as low as 4ms but the update_agent_info calls take a bit of time
+        tokio::time::sleep(Duration::from_millis(6)).await;
+
+        let called_count = test_sender.get_called_count().await.unwrap();
+        assert!(
+            called_count >= 3,
+            "Task should have run at least 3 times but was {}",
+            called_count
+        );
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn task_shuts_down_cleanly() {
+        let (test_sender, task) = setup(DummySpaceInternalImpl::new()).await;
+        test_sender.ghost_actor_shutdown().await.unwrap();
+
+        let max_wait = Instant::now();
+        while !task.read().is_finished && max_wait.elapsed() < Duration::from_millis(10) {
+            tokio::time::sleep(Duration::from_millis(1)).await;
+        }
+
+        assert!(
+            task.read().is_finished,
+            "Task should have been marked finished after the ghost actor shut down"
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn task_stays_alive_when_update_call_errors() {
+        let mut space_internal_impl = DummySpaceInternalImpl::new();
+        space_internal_impl.respond_with_error = true;
+        let (test_sender, _) = setup(space_internal_impl).await;
+
+        // It should be possible to set this as low as 4ms but the update_agent_info calls take a bit of time
+        tokio::time::sleep(Duration::from_millis(6)).await;
```

Comment:
```suggestion
        tokio::time::timeout(Duration::from_millis(100), async {
            loop {
                tokio::time::sleep(Duration::from_millis(1)).await;
                if test_sender.get_errored_count().await.unwrap() >= 3 {
                    break;
                }
            }
        }).await.unwrap();
```

Reply:
Good suggestion thanks, applied to both tests that were set up this way